python/GafferUI/CompoundEditor.py

```diff
@@ -172,40 +227,26 @@ def _removeDetachedPanel( self, panel ) :
 		panel.__removeOnCloseConnection = None
 		panel._applyVisibility()
 
+	def _preferredBound( self ) :
+
+		return self.__preferredBound
+
 	def __visibilityChanged(self, widget) :
 
 		v = self.visible()
 		for p in self._detachedPanels() :
 			p.setVisible( v )
 
-	def __serialise( self, w ) :
-
-		assert( isinstance( w, _SplitContainer ) )
-
-		if w.isSplit() :
-			sizes = w.getSizes()
-			splitPosition = ( float( sizes[0] ) / sum( sizes ) ) if sum( sizes ) else 0
-			return "( GafferUI.SplitContainer.Orientation.%s, %f, ( %s, %s ) )" % ( str( w.getOrientation() ), splitPosition, self.__serialise( w[0] ), self.__serialise( w[1] ) )
-		else :
-			# not split - a tabbed container full of editors
-			tabbedContainer = w[0]
-			tabDict = { "tabs" : tuple( tabbedContainer[:] ) }
-			if tabbedContainer.getCurrent() is not None :
-				tabDict["currentTab"] = tabbedContainer.index( tabbedContainer.getCurrent() )
-			tabDict["tabsVisible"] = tabbedContainer.getTabsVisible()
-
-			tabDict["pinned"] = []
-			for editor in tabbedContainer :
-				if isinstance( editor, GafferUI.NodeSetEditor ) :
-					tabDict["pinned"].append( not editor.getNodeSet().isSame( self.scriptNode().selection() ) )
-				else :
-					tabDict["pinned"].append( None )
-
-			return repr( tabDict )
-
 	def __repr__( self ) :
 
-		return "GafferUI.CompoundEditor( scriptNode, children = %s )" % self.__serialise( self.__splitContainer )
+		window = self.ancestor( GafferUI.ScriptWindow )
+
+		return "GafferUI.CompoundEditor( scriptNode, children = %s, detachedPanels = %s, preferredBound = %s )" \
+				% (
+					_serialiseSplits( self.__splitContainer ),
+					self._detachedPanels(),
```

Comment:
This is "leaking" the `_DetachedPanel` class into the serialisation of the layouts, but we want to keep it entirely private. Can we just serialise a dict of the arguments that we will later pass to the `_DetachedPanel` constructor?

Reply:
That would probably solve the other issues you mentioned above too...

Reply:
Fixed in https://github.com/GafferHQ/gaffer/pull/3167/commits/4361ca224e657be1db373d6d0fe91d18a252c235